syntax = "proto3";

package potato;

// ════════════════════════════════════════════════════════════
// Core Potato Domain Messages
// ════════════════════════════════════════════════════════════

// Initial seed specification - generated by AWS seed service
message PotatoSpec {
  string id = 1;                  // Unique potato identifier
  string origin = 2;              // Where seed was generated (e.g., "aws-us-east-1")
  float seed_size_mm = 3;         // Initial seed size in millimeters
  string variety = 4;             // Potato variety (russet, yukon gold, etc.)
  int64 planted_timestamp = 5;    // Unix timestamp when planted

  // Genetic lineage
  string parent_id = 6;           // Parent potato ID (empty for first generation)
  int32 generation = 7;           // Generation number (0 = original seed)
}

// Current physical growth state - updated continuously
message PotatoGrowth {
  string id = 1;

  // Physical measurements
  float mass_grams = 2;           // Current mass
  float diameter_mm = 3;          // Average diameter
  float length_mm = 4;            // Length

  // Biological properties
  float starch_content = 5;       // 0.0-1.0 (percentage)
  float water_content = 6;        // 0.0-1.0 (percentage)
  float sugar_level = 7;          // 0.0-1.0 (affects taste)

  // Growth tracking
  float growth_rate = 8;          // mm/day
  int32 days_growing = 9;         // Days since planting
  int64 last_updated = 10;        // Unix timestamp

  // Health indicators
  float health_score = 11;        // 0.0-1.0 (overall health)
  repeated string defects = 12;   // List of defects (e.g., "bruise", "blight")
}

// Genetic and visual characteristics
message PotatoDNA {
  string id = 1;

  // Visual traits
  string skin_color = 2;          // "brown", "red", "purple", "yellow"
  string flesh_color = 3;         // "white", "yellow", "purple"
  string shape = 4;               // "round", "oval", "oblong", "fingerling"
  string texture = 5;             // "smooth", "rough", "netted"

  // Genetic traits
  repeated string traits = 6;     // Special traits: "disease-resistant", "drought-tolerant", etc.
  bool has_mutation = 7;          // Whether mutation occurred
  string mutation_type = 8;       // Type of mutation (if any): "beneficial", "neutral", "detrimental"

  // Eyes and sprouts
  int32 eye_count = 9;            // Number of eyes
  float sprout_length_mm = 10;    // Average sprout length

  // Calculated metrics
  float genetic_diversity = 11;   // 0.0-1.0 (diversity score)
  float mutation_rate = 12;       // Likelihood of future mutations
}

// Aggregate potato state - the "complete picture"
message PotatoState {
  string id = 1;
  PotatoSpec spec = 2;            // Original seed info
  PotatoGrowth growth = 3;        // Current physical state
  PotatoDNA dna = 4;              // Genetic characteristics

  // Lifecycle status
  enum LifecycleStage {
    SEED = 0;
    SPROUTING = 1;
    GROWING = 2;
    MATURE = 3;
    HARVESTED = 4;
    SPOILED = 5;
  }
  LifecycleStage stage = 5;

  // Quality assessment
  enum QualityGrade {
    PREMIUM = 0;
    STANDARD = 1;
    UTILITY = 2;
    REJECT = 3;
  }
  QualityGrade grade = 6;

  // Metadata
  int64 state_version = 7;        // Incremented on each update
  string last_service = 8;        // Service that last updated this state
}

// Report message - sent to analytics/logging
message PotatoReport {
  string id = 1;
  PotatoState state = 2;

  // Processing metadata
  string service_name = 3;        // Which service generated this report
  int64 reported_at = 4;          // Unix timestamp

  // Events and observations
  repeated string events = 5;     // Events that occurred: "mutation_detected", "growth_spurt", etc.
  map<string, float> metrics = 6; // Service-specific metrics
}
